react-components/src/architecture/base/README.md

```diff
@@ -0,0 +1,186 @@
+
+- **Reuse:** I try to reuse code whenever it is possible. Therefore you will find functions that are very small,
+  often one liners. This ensures the complexity of the code to be low and increase readability on the caller side. For instance:
+
+        export function clear<T>(array: T[]): void {
+            array.splice(0, array.length);
+        }
+
+- **Single responsibility:** I try to give each class one single responsibility. Therefore I use several base classes, each class adds some complexity.
+  Examples of this is `BaseView/ThreeView/GroupThreeView` which are all base classes. If they are merge to one single class, it will be too complex.
+
+- **Interpendence:** Try to keep must classes independed of other classes. For instance the tool should not know abount the views. Utility function and classes can used whenevery you like.
+
+- **Utility function and classes:** I have added som utility function/classes in `architecture/base/utilities`.
+  - Some utility functions may be a duplicate of some function in a package we use. Please tell me.
+  - Some utility classes you will see is like a duplicate, but is made by purpose. For instance `Range3` class which is almost the same as `THREE.Box`, but is far easier to work with.
+  - Some utility classes or function is already implemented in Reveal, and are similar. The reason for not reusing Reveal here is due to what we should expose out of Reveal and into Reveral-components. For the moment this is rather strick. Examples is `Vector3Pool` which is implemented both places. Also some of the color classes may seem to be similar. But all this is low level stuff, and is not related to any Reveal functionallity.
+  - I will remove unused utility function or classes when I feel ready for it. Most of them are imported file by file from the node-visualier project.
+
+# Architecture Overview
+
+## architecture/base
+
+Here is the architecture with all base classes and some utility functionality.
+
+### architecture/base/domainObject
+
+- **DomainObject:** This is the base class for the domain object. A domain object is some sort of data, that is a collection of other domainObject or the data itself.
+
+  - An important property of the domainObject is that is has a list of views that will be updated.
+  - It has a parent and a collection of children. To ensure easy use, I have added several methods to access children, descendants and ancestors in several ways.
+
+  - It has some properties like selected, expanded and active. This is not used yet, but I wanted it to be here as this is part of the pattern. They are connected to how they are visualized and behave in a tree view, but this is not used for the moment.
```

Comment:
I imagine we will have trouble enforcing proper usage of these properties if we can't test them in any way. Could we leave them out now until we need them?